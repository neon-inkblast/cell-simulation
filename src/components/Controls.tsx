import styled from "styled-components";
import React from "react";
import { useCellGridContext } from "../context/CellGridContext";

const Button = styled.button`
  padding: 10px;
  margin: 10px;
  border-radius: 10px;
  background-color: lightgreen;
  border: 1px solid black;
`;
const ControlWrapper = styled.div`
  display: flex;
  width: 480px;
  justify-content: space-between;
  align-items: center;
  margin: 10px auto;
`;

const Controls: React.FC = () => {
  const { state, dispatch } = useCellGridContext();
  return (
    <ControlWrapper>
      <Button onClick={() => dispatch({ type: "reset" })}>Reset</Button>
      <div>Generation {state.generation}</div>
      <Button onClick={() => dispatch({ type: "generate-next" })}>Next Gen</Button>
    </ControlWrapper>
  );
};

export default Controls;
