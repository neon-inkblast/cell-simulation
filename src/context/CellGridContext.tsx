import React, { useReducer } from "react";
import { CellGridAction, CellGridContextType, CellGridState } from "../@types/CellGridContext";
import { GridConfig } from "../config/gridConfig";
import { generateNext, initialState, toggleCell } from "./cellGridUtils";

export const CellGridContext = React.createContext<CellGridContextType | undefined>(undefined);

// reducer for the state in this context
function gridReducer(state: CellGridState, action: CellGridAction): CellGridState {
  switch (action.type) {
    case "reset": {
      return initialState(GridConfig.rows, GridConfig.columns);
    }
    case "toggle-cell": {
      return toggleCell(state, action.row as number, action.column as number);
    }
    case "generate-next": {
      return generateNext(state);
    }
    default:
      throw Error(`Action type "${action.type}" not defined.  No state change`);
  }
}

// reducer wrapped provider component
export function CellGridProvider({ children }: { children: React.ReactNode }) {
  const [state, dispatch] = useReducer(gridReducer, initialState(GridConfig.rows, GridConfig.columns));
  const value = { state, dispatch };
  return <CellGridContext.Provider value={value}> {children} </CellGridContext.Provider>;
}

// hook around useContext to show an error if used outside provider
export function useCellGridContext(): CellGridContextType {
  const context = React.useContext(CellGridContext);
  if (context === undefined) {
    throw new Error("useGrid must be used within a GridProvider");
  }
  return context;
}

export default { CellGridContext, CellGridProvider, useCellGridContext };
