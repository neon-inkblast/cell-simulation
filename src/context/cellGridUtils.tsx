import { CellGridState } from "../@types/CellGridContext";
/*
 * Utility functions to perform the state transforms for the CellGridProvider
 */
export const generateNext = (state: CellGridState): CellGridState => {
  const { grid: prevGrid, generation } = state;
  // calculate max rows and columns, for cell neighbour wrapping
  const lastRowIndex = prevGrid.length ? prevGrid.length - 1 : 0;
  const lastColIndex = prevGrid[0]?.length ? prevGrid[0].length - 1 : 0;

  // inner function to use when mapping current cells to next gen cells
  function getNextCellState(x: number, y: number, cell: boolean): boolean {
    // calculate surrounding rows and columns, taking wrapping into account
    const prevRow = y === 0 ? lastRowIndex : y - 1;
    const nextRow = y === lastColIndex ? 0 : y + 1;
    const prevCol = x === 0 ? lastRowIndex : x - 1;
    const nextCol = x === lastColIndex ? 0 : x + 1;

    // get an array of the neighbouring cells
    const neighbours = [
      prevGrid[prevRow][prevCol],
      prevGrid[prevRow][x],
      prevGrid[prevRow][nextCol],
      prevGrid[y][prevCol],
      prevGrid[y][nextCol],
      prevGrid[nextRow][prevCol],
      prevGrid[nextRow][x],
      prevGrid[nextRow][nextCol],
    ];

    // count the "live" neighbouring cells
    const liveNeighbours = neighbours.reduce((acc, current) => (acc += current ? 1 : 0), 0);
    // if live neighbours is 3, become or stay alive
    // otherwise, if 2 and alive, stay alive
    return liveNeighbours === 3 || (liveNeighbours === 2 && cell);
  }

  // map the new cell states from the old cells
  const grid = prevGrid.map((row, y) =>
    row.map((cell, x) => {
      return getNextCellState(x, y, cell);
    })
  );

  return {
    generation: generation + 1,
    grid,
  };
};

export const toggleCell = (state: CellGridState, cellY: number, cellX: number): CellGridState => {
  const prevGrid = state.grid;
  const grid = prevGrid.map((row, y) => {
    // if not the target row, skip any more checks and just recreate the same row
    if (y !== cellY) {
      return [...row];
    }
    // if in the target row, toggle when the x position matches
    return row.map((cell, x) => {
      if (cellX === x) {
        return !cell;
      }
      return cell;
    });
  });
  return {
    ...state,
    grid,
  };
};

export const initialState = (rows = 5, columns = 5): CellGridState => {
  return {
    generation: 0,
    grid: Array(rows).fill(Array(columns).fill(false)),
  };
};

const cellGridUtils = { generateNext, toggleCell, initialState };

export default cellGridUtils;
