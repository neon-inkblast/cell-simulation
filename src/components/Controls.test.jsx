import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { CellGridContext } from "../context/CellGridContext";
import Controls from "./Controls";

const customRender = (ui, { providerProps, ...renderOptions }) => {
  return render(<CellGridContext.Provider value={{ ...providerProps }}>{ui}</CellGridContext.Provider>, renderOptions);
};

describe("Controls component", () => {
  let renderResult;
  let providerProps;

  beforeEach(() => {
    providerProps = {
      state: {
        generation: 0,
        grid: [
          [0, 1],
          [1, 0],
        ],
      },
      dispatch: jest.fn(),
    };
    renderResult = customRender(<Controls />, { providerProps });
  });

  it("renders", () => {
    expect(renderResult.asFragment()).toMatchSnapshot();
  });

  it("dispatches a reset event on Reset button click", () => {
    // Click the reset button
    const resetButton = screen.getByText(/reset/i);
    userEvent.click(resetButton);
    // Verify dispatch to context state was called
    expect(providerProps.dispatch).toHaveBeenCalledTimes(1);
    expect(providerProps.dispatch).toHaveBeenCalledWith({ type: "reset" });
  });

  it("dispatches a next generation event on Next button click", () => {
    // Click the next button
    const nextButton = screen.getByText(/next/i);
    userEvent.click(nextButton);
    // Verify dispatch to context state was called
    expect(providerProps.dispatch).toHaveBeenCalledTimes(1);
    expect(providerProps.dispatch).toHaveBeenCalledWith({ type: "generate-next" });
  });
});
