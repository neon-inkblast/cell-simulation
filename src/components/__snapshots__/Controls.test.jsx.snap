// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Controls component renders 1`] = `
<DocumentFragment>
  <div
    class="sc-gsTEea fxyNSE"
  >
    <button
      class="sc-bdfBQB icJfVi"
    >
      Reset
    </button>
    <button
      class="sc-bdfBQB icJfVi"
    >
      Next Gen
    </button>
    <div>
      Generation 0
    </div>
  </div>
</DocumentFragment>
`;
